### PR TITLE
feat(Kafka): Add option to login to SASL password-protected Kafka, like MSK

### PR DESCRIPTION
**Pull request by Smejky338**

### CODE REVIEW

**Comment by Smejky338** [comment]
In the first use, I got these timeout errors so that only a part of the hosts generated:
https://master-jenkins-csb-perf.apps.ocp-c1.prod.psi.redhat.com/job/InsightsROS_runner/397/console
`2023-12-08 15:44:12,709 kafka.client WARNING <BrokerConnection node_id=3 host=b-3-public.openpocmsk.pyzafp.c6.kafka.us-east-2.amazonaws.com:9196 <connected> [IPv4 ('3.132.12.218', 9196)]> timed out after 30000 ms. Closing connection.
2023-12-08 15:44:12,709 kafka.client WARNING Node 3 connection failed -- refreshing metadata
2023-12-08 15:44:12,710 kafka.producer.record_accumulator WARNING Produced messages to topic-partition TopicPartition(topic='platform.inventory.host-ingress', partition=8) with base offset -1 log start offset None and error None.
2023-12-08 15:44:12,710 root ERROR Failed to send message (RequestTimedOutError('Request timed out after 30000 ms'),); {'data_stats': {'successes': 7, 'failures': 0}, 'data_lock': <unlocked _thread.lock object at 0x7f44f173ccc0>}
2023-12-08 15:44:12,710 kafka.client WARNING <BrokerConnection node_id=2 host=b-2-public.openpocmsk.pyzafp.c6.kafka.us-east-2.amazonaws.com:9196 <connected> [IPv4 ('3.131.20.190', 9196)]> timed out after 30000 ms. Closing connection.
2023-12-08 15:44:12,710 kafka.producer.record_accumulator WARNING [Error 7] RequestTimedOutError: Request timed out after 30000 ms
2023-12-08 15:44:12,710 kafka.client WARNING Node 2 connection failed -- refreshing metadata
2023-12-08 15:44:12,711 kafka.producer.record_accumulator WARNING Produced messages to topic-partition TopicPartition(topic='platform.inventory.host-ingress', partition=1) with base offset -1 log start offset None and error None.
2023-12-08 15:44:12,711 root ERROR Failed to send message (RequestTimedOutError('Request timed out after 30000 ms'),); {'data_stats': {'successes': 7, 'failures': 1}, 'data_lock': <unlocked _thread.lock object at 0x7f44f173ccc0>}
2023-12-08 15:44:12,711 kafka.producer.record_accumulator WARNING Produced messages to topic-partition TopicPartition(topic='platform.inventory.host-ingress', partition=4) with base offset -1 log start offset None and error None.
2023-12-08 15:44:12,711 root ERROR Failed to send message (RequestTimedOutError('Request timed out after 30000 ms'),); {'data_stats': {'successes': 7, 'failures': 2}, 'data_lock': <unlocked _thread.lock object at 0x7f44f173ccc0>}
2023-12-08 15:44:12,711 kafka.producer.record_accumulator WARNING [Error 7] RequestTimedOutError: Request timed out after 30000 ms
2023-12-08 15:44:12,711 kafka.producer.record_accumulator WARNING [Error 7] RequestTimedOutError: Request timed out after 30000 ms
Traceback (most recent call last):
  File "/home/venv/bin/script-hbi-populate.py", line 33, in <module>
    sys.exit(load_entry_point('opl-rhcloud-perf-team', 'console_scripts', 'script-hbi-populate.py')())
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/venv/src/opl-rhcloud-perf-team/opl/hbi_utils.py", line 311, in populate_main
    gen_send_verify(args, status_data)
  File "/home/venv/src/opl-rhcloud-perf-team/opl/hbi_utils.py", line 217, in gen_send_verify
    gen_and_send(
  File "/home/venv/src/opl-rhcloud-perf-team/opl/hbi_utils.py", line 107, in gen_and_send
    raise Exception(
Exception: Failed to send 3 messages out of totally requested 10
command terminated with exit code 1`

... but the second one worked fine: https://master-jenkins-csb-perf.apps.ocp-c1.prod.psi.redhat.com/job/InsightsROS_runner/398/console

File: opl/consumer_lag.py [hunk 1]
```diff
@@ -12,26 +12,52 @@ class ConsumerLag:
...
-            max_poll_records=50,
-            max_poll_interval_ms=300000,
-            group_id=self.group,
-            session_timeout_ms=50000,
-            heartbeat_interval_ms=10000,
-            consumer_timeout_ms=100000,
-        )
+        # Common parameters for both cases
+        common_params = {
+            topic: self.topic,
```
Comment (shubham-html-css-js) [inline opl/consumer_lag.py]: I think this "topic" key should be removed from here too

File: opl/consumer_lag.py [hunk 2]
```diff
@@ -12,26 +12,52 @@ class ConsumerLag:
...
+            bootstrap_servers: self.bootstrap_servers,
+            auto_offset_reset: "latest",
+            enable_auto_commit: False,
+            max_poll_records: 50,
+            max_poll_interval_ms: 300000,
+            group_id: self.group,
+            session_timeout_ms: 50000,
+            heartbeat_interval_ms: 10000,
+            consumer_timeout_ms: 100000,
```
Comment (shubham-html-css-js) [inline opl/consumer_lag.py]: I think the keys should be wrapped in quotes

File: opl/skip_to_end.py [hunk 3]
```diff
@@ -20,20 +20,39 @@ def doit_seek_to_end(kafka_hosts, kafka_timeout, kafka_topic, kafka_group):
...
 
+    # Common parameters for both cases
+    common_params = {
+        bootstrap_servers: kafka_hosts,
+        auto_offset_reset: "latest",
+        enable_auto_commit: True,
+        group_id: kafka_group,
+        session_timeout_ms: 50000,
+        heartbeat_interval_ms: 10000,
+        consumer_timeout_ms: kafka_timeout,
```
Comment (shubham-html-css-js) [inline opl/skip_to_end.py]: I think the keys should be wrapped in quotes

**Review by shubham-html-css-js — CHANGES_REQUESTED** [review]

File: opl/consumer_lag.py (hunk 2 above)
Comment (Smejky338) [reply opl/consumer_lag.py:38]: I see, it shouldn't subscribe there anyways as it uses this method: https://kafka-python.readthedocs.io/en/master/apidoc/KafkaConsumer.html?highlight=partitions_for_topic#kafka.KafkaConsumer.partitions_for_topic

File: opl/consumer_lag.py (hunk 1 above)
Comment (Smejky338) [reply opl/consumer_lag.py]: I see, it shouldn't subscribe there anyways as it uses this method: https://kafka-python.readthedocs.io/en/master/apidoc/KafkaConsumer.html?highlight=partitions_for_topic#kafka.KafkaConsumer.partitions_for_topic

File: opl/consumer_lag.py [hunk 4]
```diff
@@ -12,26 +12,50 @@ class ConsumerLag:
...
-            enable_auto_commit=False,
-            max_poll_records=50,
-            max_poll_interval_ms=300000,
-            group_id=self.group,
-            session_timeout_ms=50000,
-            heartbeat_interval_ms=10000,
-            consumer_timeout_ms=100000,
-        )
+        # Common parameters for both cases
+        common_params = {
```
Comment (jhutar) [inline opl/consumer_lag.py:28]: Nice!

File: opl/consumer_lag.py [hunk 5]
```diff
@@ -12,26 +12,50 @@ class ConsumerLag:
...
+            "max_poll_records": 50,
+            "max_poll_interval_ms": 300000,
+            "group_id": self.group,
+            "session_timeout_ms": 50000,
+            "heartbeat_interval_ms": 10000,
+            "consumer_timeout_ms": 100000,
+        }
 
+        # Kafka consumer creation: SASL or noauth
+        try:
```
Comment (jhutar) [inline opl/consumer_lag.py]: With this try we will print first log info message every time. Could you replace the `try:` directly with `if self.username != "":`?

Comment (jhutar) [reply opl/consumer_lag.py]: Or maybe `if self.username is not None:` given how default is specified in `add_kafka_opts(`)

File: opl/get_kafka_times.py [hunk 6]
```diff
@@ -78,26 +80,43 @@ def kafka_ts2dt(self, timestamp):
...
-        logging.debug(
-            f"Created Kafka consumer for {self.kafka_hosts} for {self.kafka_topic} topic in group {self.kafka_group} with {self.kafka_timeout} ms timeout"
-        )
+        try:
+            logging.info(
+                f"Creating consumer with sasl username&pasword to {self.kafka_host}"
+            )
+            consumer = KafkaConsumer(
+                self.kafka_topic,
+                bootstrap_servers=self.kafka_host,  # self.kafka_hosts,
```
Comment (jhutar) [inline opl/get_kafka_times.py]: I think `bootstrap_servers` needs to be a list, not just one string.

File: opl/get_kafka_times.py [hunk 7]
```diff
@@ -78,26 +80,43 @@ def kafka_ts2dt(self, timestamp):
...
-            group_id=self.kafka_group,
-            max_poll_records=self.kafka_max_poll_records,
-            session_timeout_ms=50000,
-            heartbeat_interval_ms=10000,
-            consumer_timeout_ms=self.kafka_timeout,
-        )
-        logging.debug(
-            f"Created Kafka consumer for {self.kafka_hosts} for {self.kafka_topic} topic in group {self.kafka_group} with {self.kafka_timeout} ms timeout"
-        )
+        try:
```
Comment (jhutar) [inline opl/get_kafka_times.py]: Same here. Can you do a common list of options and only add sasl and security opts `if username is not None`?

File: opl/hbi_utils.py [hunk 8]
```diff
@@ -188,10 +188,26 @@ def gen_send_verify(args, status_data):
     if args.dry_run:
         producer = None
     else:
+        try:
```
Comment (jhutar) [inline opl/hbi_utils.py]: Again, please use if so we do not log `Creating SASL password-protected producer to {args.kafka_host}` everytime.

File: opl/post_kafka_times.py [hunk 9]
```diff
@@ -328,18 +328,35 @@ def produce_thread(args, config, produce_here, save_here):
...
+            "retries": args.retries,
+            "batch_size": args.batch_size,
+            "buffer_memory": args.buffer_memory,
+            "linger_ms": args.linger_ms,
+            "max_block_ms": args.max_block_ms,
+            "request_timeout_ms": args.request_timeout_ms,
+            "compression_type": args.compression_type,
+        }
+
+        try:
```
Comment (jhutar) [inline opl/post_kafka_times.py]: Also here please.

File: opl/skip_to_end.py [hunk 10]
```diff
@@ -20,20 +20,39 @@ def doit_seek_to_end(kafka_hosts, kafka_timeout, kafka_topic, kafka_group):
...
+        "auto_offset_reset": "latest",
+        "enable_auto_commit": True,
+        "group_id": kafka_group,
+        "session_timeout_ms": 50000,
+        "heartbeat_interval_ms": 10000,
+        "consumer_timeout_ms": kafka_timeout,
+    }
+
+    # Kafka consumer creation: SASL or noauth
+    try:
```
Comment (jhutar) [inline opl/skip_to_end.py]: You know :)

File: opl/skip_to_end.py [hunk 11]
```diff
@@ -57,6 +76,7 @@ def doit(args, status_data):
         args.kafka_timeout,
         args.kafka_topic,
         args.kafka_group,
+        args,
```
Comment (jhutar) [inline opl/skip_to_end.py]: Change it just to send args then as we have all of the other stuff there anyway

**Review by shubham-html-css-js — APPROVED** [review]

**Comment by shubham-html-css-js** [comment]
@jsmejkal @jhutar I have tried this in my local environment and for Patchman , the script works well

File: opl/consumer_lag.py (hunk 5 above)
Comment (Smejky338) [reply opl/consumer_lag.py]: Or maybe I should set the add_kafka_opts default to ""?

File: opl/get_kafka_times.py (hunk 6 above)
Comment (Smejky338) [reply opl/get_kafka_times.py]: It works even as a single string: https://kafka-python.readthedocs.io/en/master/apidoc/KafkaConsumer.html
This way should be more flexible to alter between single and many servers, but I guess the flexible way won't work without some additional code :)

File: opl/skip_to_end.py (hunk 11 above)
Comment (Smejky338) [reply opl/skip_to_end.py]: Good point!

**Comment by Smejky338** [comment]
I've changed to default the kafka_user and kafka_password to "" and modified the scripts to take that into account and also refactored some Kafka consumer/producer code to be separated into a method.
It would be nice to refactor all the kafka consumer and producer methods into a common place so we don't have that much copypaste, but I would skip that "for now".

**Review by jhutar — APPROVED** [review]

**Review by jhutar — APPROVED** [review]

**Pull request merged**